File: FitBuddy/MainViewControllers/LogbookViewController.m

//
//  LogbookViewController.m
//  GymBuddy
//
//  Created by John Neyer on 2/12/12.
//  Copyright (c) 2012 jneyer.com. All rights reserved.
//

#import "LogbookViewController.h"
#import "FitBuddyMacros.h"
#import "BarChartDataSource.h"
#import "JBChartView.h"

#import "FitBuddy-Swift.h"

@implementation LogbookViewController
{
    JBBarChartView *chart;
    NSFetchedResultsController *chartDataSource;
    NSMutableArray *entries;
    BarChartDataSource *chartData;
    BOOL frameLoaded;
}

@synthesize logbookEntry = _logbookEntry;

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.navigationItem.titleView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:kFITBUDDY]];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(setupFetchedResultsController) name:kUBIQUITYCHANGED object:nil];
    
    entries = [[NSMutableArray alloc]init];
    
    
}

-(void) setupFetchedResultsController
{
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:LOGBOOK_TABLE];
    request.sortDescriptors = [NSArray arrayWithObject:[NSSortDescriptor sortDescriptorWithKey:@"date" ascending:NO selector:@selector(compare:)]];
    request.predicate = [NSPredicate predicateWithFormat:@"completed = %@", [NSNumber numberWithBool:YES]];

    self.fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:request 
                                                                        managedObjectContext: [[AppDelegate sharedAppDelegate]managedObjectContext]
                                                                          sectionNameKeyPath:@"date_t"
                                                                                   cacheName:nil];
    
    
}

-(void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self setupFetchedResultsController];
}

-(void) viewDidAppear:(BOOL)animated
{
    if (chartData == nil)
    {
        chartData = [[BarChartDataSource alloc] init];
    }
    
    [chartData load];
    [self prepareChart];
    [chart reloadData];
    
    [super viewDidAppear:animated];

    CGRect footerframe = chart.footerView.frame;
    footerframe.origin.y = footerframe.origin.y + 20;
    [chart.footerView setFrame: footerframe];
    
}

- (void) prepareChart
{
    if (chart == nil)
    {
        chart = [[JBBarChartView alloc]init];
        [chart setDelegate:chartData];
        [chart setDataSource:chartData];
        
        [self addSubview:chart fillingAndInsertedIntoView:self.chartView];
    
        UILabel *headerView = [[UILabel alloc] initWithFrame:CGRectMake(10.0f, ceil(chart.bounds.size.height * 0.5) - ceil(80.0f * 0.5), chart.bounds.size.width - (10.0f * 2), 80.0f)];
        headerView = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 200, 14)];
        
        [headerView setText: [@"30 Day Activity Profile" uppercaseString]];
        [headerView setTextAlignment:NSTextAlignmentCenter];
        [headerView setTextColor: kCOLOR_DKGRAY];
        [headerView setFont:[UIFont fontWithName:headerView.font.fontName size:14.0f]];
        chart.headerView = headerView;
        
        [chart setHeaderPadding:10.0f];
        
        NSDateFormatter *dateFormat = [[NSDateFormatter alloc] init];
        [dateFormat setDateFormat:@"dd MMM yyyy"];
        
        NSString *rightDate = [dateFormat stringFromDate:[[NSDate alloc] init]];
        
        UILabel *rightLabel = [[UILabel alloc] initWithFrame:CGRectMake(20.0f, 400, chart.bounds.size.width - (10.0f * 2), 40.0f)];
        rightLabel = [[UILabel alloc] initWithFrame:CGRectMake(0,0,200,12)];
        [rightLabel setText: [rightDate uppercaseString]];
        [rightLabel setTextAlignment:NSTextAlignmentRight];
        [rightLabel setTextColor: [UIColor blackColor]];
        [rightLabel setFont:[UIFont fontWithName:rightLabel.font.fontName size:12.0f]];
        chart.footerView = rightLabel;
        
        [chart layoutIfNeeded];

    }
}


- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 60.0;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 40.0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Logbook Cell"];
    UILabel *exerciseLabel = (UILabel *)[cell viewWithTag:100];
    UILabel *workoutLabel = (UILabel *) [cell viewWithTag:101];
    UIImageView *exerciseIcon = (UIImageView *) [cell viewWithTag:102];
    UILabel *qstatValue = (UILabel *) [cell viewWithTag:103];
    UILabel *qstatLabel = (UILabel *) [cell viewWithTag:104];
    
    // Add the data to the cell
    LogbookEntry *logbookEntry = [self.fetchedResultsController objectAtIndexPath:indexPath];
    [entries addObject:logbookEntry];

    exerciseLabel.text = logbookEntry.exercise_name;
    workoutLabel.text = logbookEntry.workout_name;
    
    if (logbookEntry.pace || logbookEntry.distance || logbookEntry.duration)
    {
        exerciseIcon.image = [UIImage imageNamed:kCARDIO];
        qstatLabel.text = @"Pace";
        qstatValue.text = logbookEntry.pace;
    }
    else
    {
        exerciseIcon.image = [UIImage imageNamed:kRESISTANCE];
        qstatLabel.text = @"Weight";
        qstatValue.text = logbookEntry.weight;
    }
    
    return cell;
    
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    //if (DEBUG) NSLog(@"Segue: %@", segue.identifier);
    
    LogbookEntry *entry = [self.fetchedResultsController objectAtIndexPath:[self.tableView indexPathForSelectedRow]];
    if ([segue.destinationViewController respondsToSelector:@selector(setLogbookEntry:)]) {
        [segue.destinationViewController performSelector:@selector(setLogbookEntry:) withObject:entry];
    }

}


- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath 
{
    // Return YES to edit
    return YES;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [self.tableView reloadData];
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
  
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) 
    {
        //Check if index path is valid
        if(indexPath)
        {
            //Get the cell out of the table view
            UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
            
            //Update the cell or model 
            cell.editing = YES;
            LogbookEntry *entry = [self.fetchedResultsController objectAtIndexPath:indexPath];
            [[AppDelegate sharedAppDelegate].managedObjectContext deleteObject:entry];
        }
    }    
}

-(NSString *) convertRawToShortDateString: (NSString *) rawDateStr
{
    // Convert rawDateStr string to NSDate...
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyy-MM-dd HH:mm:ss ZZ"];
    NSDate *date = [formatter dateFromString:rawDateStr];
    
    // Convert NSDate to format we want...
    [formatter setDateFormat:@"dd MMMM yyyy"];
    NSString *formattedDateStr = [formatter stringFromDate:date];
    return formattedDateStr;  
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    
    NSString *sectionTitle = [self tableView:tableView titleForHeaderInSection:section];
    if (sectionTitle == nil) {
        return nil;
    }
    
    UIView *labelView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, tableView.frame.size.width, 40.0)];
    [labelView setBackgroundColor: kCOLOR_LTGRAY];
    [labelView setAutoresizesSubviews:TRUE];
    
    // Create label with section title
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(15, 5, tableView.frame.size.width, 40.0)];
    label.font = [UIFont systemFontOfSize:14.0];
    label.text = [[self convertRawToShortDateString:sectionTitle] capitalizedString];
    [label setTextColor: kCOLOR_DKGRAY];

    [labelView addSubview:label];
    
    return labelView;
}

- (NSArray *)sectionIndexTitlesForTableView:(UITableView *)tableView {
    return nil;
}

- (NSInteger)tableView:(UITableView *)tableView sectionForSectionIndexTitle:(NSString *)title atIndex:(NSInteger)index {
    return 0;
}

#pragma mark - JBChartView Delegate
- (void)addSubview:(UIView *)insertedView fillingAndInsertedIntoView:(UIView *)containerView {
    
    [containerView addSubview:insertedView];
    [insertedView setTranslatesAutoresizingMaskIntoConstraints:NO];
    
    [containerView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[insertedView]|"
                                                                          options:0
                                                                          metrics:nil
                                                                            views:NSDictionaryOfVariableBindings(insertedView)]];
    [containerView addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[insertedView]|"
                                                                          options:0
                                                                          metrics:nil
                                                                            views:NSDictionaryOfVariableBindings(insertedView)]];
    
    [containerView layoutIfNeeded];
}


@end

File: controllers/LogbookViewController.swift
//
//  LogbookViewController.swift
//  FitBuddy
//
//  Created by john.neyer on 5/16/15.
//  Copyright (c) 2015 jneyer.com. All rights reserved.
//

import Foundation
import UIKit
import FitBuddyModel
import FitBuddyCommon

class LogbookViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    @IBOutlet weak var chartView: UIView!
    @IBOutlet weak var logbookTableView: UITableView!
    
    //Logbook array data
    var tableData : [LogbookEntry]?
    
    let logbookCellIdentifier = "Exercise Cell"
    let workoutCellIdentifier = "Workout Cell"
    
    var tableStyle = LogbookStyle.WORKOUT
    
    var chartLabels = NSMutableArray()
    
    let chartData = LogbookChartData()
    let logbookTableData = LogbookTableData()
    
    var workoutSection = 0
    var workoutIndex = 0
    
    lazy var chart : LineChart = {
        var chart = LineChart()
        chart.y.labels.visible = false
        chart.y.axis.visible = false
        chart.x.labels.visible = true
        chart.frame = self.chartView.frame
        chart.frame.origin.x = 0
        chart.area = false
        chart.x.grid.visible = false
        chart.y.grid.visible = false
        chart.showZeros = false
        chart.lineWidth = 0.0
        self.chartView.addSubview(chart)
        
        self.chartView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:[insertedView]|", options:NSLayoutFormatOptions.DirectionLeftToRight ,metrics: nil, views: ["insertedView": chart]))
        self.chartView.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:[insertedView]|", options:NSLayoutFormatOptions.DirectionLeftToRight ,metrics: nil, views: ["insertedView": chart]))
        
        self.chartView.layoutIfNeeded()
        
        return chart
    }()
    
    override func viewDidLoad() {
        self.navigationItem.titleView = UIImageView(image: UIImage(named:FBConstants.kFITBUDDY))
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "setupFetchedResultsController", name: FBConstants.kUBIQUITYCHANGED, object: nil)
        loadData()
    }

    override func viewWillAppear(animated: Bool) {
        
        // Initialize the chart
        chart.clearAll()

        //Set chart values
        self.chartLabels = NSMutableArray(array: chartData.lastThirty())
    
        let first = self.chartLabels.firstObject as! String
        
        //remove every fifth element day string
        for index in 1...self.chartLabels.count {
            if index % 5 != 0 {
                self.chartLabels[self.chartLabels.count - 1 - index] =  ""
            }
        }
        
        self.chartLabels[0] = first
        
        chart.x.labels.values = self.chartLabels as [AnyObject] as! [String]
        chart.addLine(chartData.normalizedResistanceArray() as [AnyObject] as! [CGFloat])
        chart.addLine(chartData.normalizedCardioArray() as [AnyObject] as! [CGFloat])
        
        loadData()
    }
    
    override func viewDidAppear(animated: Bool) {
        loadData()
        logbookTableView.reloadData()
    }
    
    func loadData() {
        tableData = AppDelegate.sharedAppDelegate().modelManager.getAllLogbookEntries()
        chartData.setLogbookData(tableData!)
        logbookTableData.setData(tableData!)
    }
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        
        if tableStyle == LogbookStyle.WORKOUT {
            return logbookTableData.numberOfSections()
        }
        
        if tableStyle == LogbookStyle.EXERCISE {
            return 1
        }
        
        return 0
    }

    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        if tableStyle == LogbookStyle.WORKOUT {
            return logbookTableData.numberOfRowsInSection(section)
        }
            
        if  tableStyle == LogbookStyle.EXERCISE {
            return logbookTableData.numberOfExercisesInWorkout(workoutSection, index: workoutIndex)
        }
        
        return 0
    }
    
    func tableView(tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        
        var text = ""
        
        if tableStyle == LogbookStyle.WORKOUT {
            text = logbookTableData.sectionAtIndex(section)
        }
        
        if tableStyle == LogbookStyle.EXERCISE {
            text = logbookTableData.sectionAtIndex(workoutSection)
        }
        
        let labelView = UIView(frame: CGRectMake(0, 0, tableView.frame.size.width, 35.0))
        labelView.backgroundColor = FBConstants.kCOLOR_LTGRAY
        labelView.autoresizesSubviews = false
        
        let label = UILabel(frame: CGRectMake(15, 0, tableView.frame.size.width, 35.0))
        label.font = UIFont.systemFontOfSize(12.0)
        label.text = text.uppercaseString
        label.textColor = FBConstants.kCOLOR_DKGRAY
        
        labelView.addSubview(label)
        return labelView
        
    }
    
    func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 35.0
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        if tableStyle == LogbookStyle.EXERCISE {
            return exerciseCellAtIndexPath(indexPath)
        }
        
        return workoutCellAtIndexPath(indexPath)
    }
    
    func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        
        if tableStyle == LogbookStyle.EXERCISE {
            return true
        }
        
        return false
    }
    
    func sectionIndexTitlesForTableView(tableView: UITableView) -> [AnyObject]! {
        return nil
    }
    
    func tableView(tableView: UITableView, sectionForSectionIndexTitle title: String, atIndex index: Int) -> Int {
        return 0
    }
    
    func exerciseCellAtIndexPath (indexPath:NSIndexPath) -> ExerciseLogCell {
        
        let cell = logbookTableView.dequeueReusableCellWithIdentifier(logbookCellIdentifier) as! ExerciseLogCell
    
        let entry = logbookTableData.exerciseForWorktoutAtIndex(workoutSection, workoutIndex: workoutIndex, exerciseIndex: indexPath.row)
        
        if entry.distance != nil {
            cell.setCellValues(name: entry.exercise_name, workout: entry.workout_name, value: entry.distance, valueType: "distance", exerciseType: ExerciseType.CARDIO)
        } else {
            cell.setCellValues(name: entry.exercise_name, workout: entry.workout_name, value: entry.weight, valueType: "weight", exerciseType: ExerciseType.RESISTANCE)
        }
        
        return cell
    }
    
    func workoutCellAtIndexPath (indexPath:NSIndexPath) -> WorkoutLogCell {
    
        let cell = logbookTableView.dequeueReusableCellWithIdentifier(workoutCellIdentifier) as! WorkoutLogCell
        let value = logbookTableData.workoutAtIndex(indexPath.section, index: indexPath.row)
        cell.setCellValues(name: value.name, date: value.date)
        
        return cell
    }
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 50.0
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        self.logbookTableView.deselectRowAtIndexPath(indexPath, animated: true)
        if tableStyle == LogbookStyle.WORKOUT {
            self.workoutSection = indexPath.section
            self.workoutIndex = indexPath.row
            self.performSegueWithIdentifier("WorkoutDetailsSeque", sender: self)
        }
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "WorkoutDetailsSeque" {
            let controller = segue.destinationViewController as! LogbookViewController
            controller.title = "Exercises"
            controller.tableStyle = LogbookStyle.EXERCISE
            controller.workoutSection = self.workoutSection
            controller.workoutIndex = self.workoutIndex
        }
        else {
            self.title = "Workouts"
        }
    }
    
    func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        
        if editingStyle == UITableViewCellEditingStyle.Delete
        {
            if let cell = tableView.cellForRowAtIndexPath(indexPath) {
                let numberOfRows = logbookTableData.numberOfExercisesInWorkout(workoutSection, index: workoutIndex)
                cell.editing = true
                let entry = logbookTableData.exerciseForWorktoutAtIndex(workoutSection, workoutIndex: workoutIndex, exerciseIndex: indexPath.row)
                AppDelegate.sharedAppDelegate().modelManager.deleteDataObject(entry)
                
                //if this was the last row pop to root
                if numberOfRows == 1 {
                    self.navigationController?.popToRootViewControllerAnimated(true)
                }
                
                loadData()
            }
            tableView.reloadData()
        }
    }
}

enum LogbookStyle {
    case WORKOUT
    case EXERCISE
}

